pytorch_lightning/core/hooks.py

```diff
@@ -471,7 +472,7 @@ def train_dataloader(self):
                 # each batch will be a dict of tensors: {'mnist': batch_mnist, 'cifar': batch_cifar}
                 return {'mnist': mnist_loader, 'cifar': cifar_loader}
         """
-        raise NotImplementedError("`train_dataloader` must be implemented to be used with the Lightning Trainer")
```

Comment:
I don't get why PyCharm shows that warning. Other classes in our codebase raise `NotImplementedError` but they don't have that warning. For example:

https://github.com/PyTorchLightning/pytorch-lightning/blob/23450e29055addcf8f1100b4fc3a4ed4fa9c0e2d/pytorch_lightning/overrides/base.py#L39

https://github.com/PyTorchLightning/pytorch-lightning/blob/23450e29055addcf8f1100b4fc3a4ed4fa9c0e2d/pytorch_lightning/callbacks/progress/base.py#L140

(among others)

Reply:
Yes, I also find it rude. If this changes in the future, we can change it back too.

Reply:
minimal repro:

```python
class Foo:
    def fn(self):
        raise NotImplementedError


class HasWarning(Foo):
    ...


class NoWarning:
    def fn(self):
        raise NotImplementedError
```

This looks like a false positive. Already reported in https://youtrack.jetbrains.com/issue/PY-38680

We can just disable the warning

Reply:
do you recommend to close the PR? is it important that we raise NotImplemented?
I read the [docs](https://docs.python.org/3/library/exceptions.html#NotImplementedError) and interpret it differently

Reply:
We can close this or add `# noqa`

Unfortunately, there's no mechanism to silence that warning in particular so any other warnings would be hidden.

I personally lean towards closing this but no strong opinion.

Reply:
I tried adding noqa to the lines where we raise NotImiplementedError and the signature of the method but both did not work

Reply:
`class LightningModule(  # noqa` works for me

No need to add it to the `NotImplementedError`

Reply:
If adding the noqa marker, could we link to the JetBrains issue in the code to know why it's there? 

Once that issue is fixed I'd prefer using NotImplementedError as its canonical Python while Misconfiguration is Lightning specific. It's very minor though, as long as we raise an exception compared to before of logging a warning message.

Reply:
@carmocca any adding noqa to the base class gets rid of the warning there but not in any of the derived classes.